rust/kernel/net/socket/opts.rs

```diff
@@ -0,0 +1,1222 @@
+
+use kernel::bindings;
+
+/// Options level to retrieve and set socket options.
+/// See `man 7 socket` for more information.
+#[derive(Debug, Copy, Clone, PartialEq, Eq, Hash)]
+pub enum OptionsLevel {
+    /// IP level socket options.
+    /// See `man 7 ip` for more information.
+    Ip = bindings::IPPROTO_IP as isize,
+
+    /// Socket level socket options.
+    /// See `man 7 socket` for more information.
+    Socket = bindings::SOL_SOCKET as isize,
+
+    /// IPv6 level socket options.
+    /// See `man 7 ipv6` for more information.
+    Ipv6 = bindings::IPPROTO_IPV6 as isize,
+
+    /// Raw level socket options.
+    /// See `man 7 raw` for more information.
+    Raw = bindings::IPPROTO_RAW as isize,
+
+    /// TCP level socket options.
+    /// See `man 7 tcp` for more information.
+    Tcp = bindings::IPPROTO_TCP as isize,
+}
+
+/// Generic socket option type.
+///
+/// This trait is implemented by each individual socket option.
+///
+/// Having socket options as structs instead of enums allows:
+/// - Type safety, making sure that the correct type is used for each option.
+/// - Read/write enforcement, making sure that only readable options
+/// are read and only writable options are written.
+pub trait SocketOption {
+    /// Rust type of the option value.
+    ///
+    /// This type is used to store the value of the option.
+    /// It is also used to enforce type safety.
+    ///
+    /// For example, the [`ip::Mtu`] option has a value of type `u32`.
+    type Type;
+
+    /// Retrieve the C value of the option.
+    ///
+    /// This value is used to pass the option to the kernel.
+    fn value() -> isize;
+
+    /// Retrieve the level of the option.
+    ///
+    /// This value is used to pass the option to the kernel.
+    fn level() -> OptionsLevel;
+}
+
+/// Generic readable socket option type.
+///
+/// This trait is implemented by each individual readable socket option.
+/// Can be combined with [`WritableOption`] to create a readable and writable socket option.
+pub trait WritableOption: SocketOption {}
+
+/// Generic writable socket option type.
+///
+/// This trait is implemented by each individual writable socket option.
+/// Can be combined with [`ReadableOption`] to create a readable and writable socket option.
+pub trait ReadableOption: SocketOption {}
+
+/// Generates the code for the implementation of a socket option.
+///
+/// # Parameters
+/// * `$opt`: Name of the socket option.
+/// * `$value`: C value of the socket option.
+/// * `$level`: Level of the socket option, like [`OptionsLevel::Ip`].
+/// * `$rtyp`: Rust type of the socket option.
+/// * `$($tr:ty),*`: Traits that the socket option implements, like [`WritableOption`].
+macro_rules! impl_opt {
+    ($(#[$meta:meta])*
+    $opt:ident = $value:expr,
+    $level:expr,
+    unimplemented,
+    $($tr:ty),*) => {};
+
+    ($(#[$meta:meta])*
+    $opt:ident = $value:expr,
+    $level:expr,
+    $rtyp:ty,
+    $($tr:ty),*) => {
```

Comment:
```suggestion
    (
        $(#[$meta:meta])*
        $opt:ident = $value:expr,
        $level:expr,
        unimplemented,
        $($tr:ty),*
    ) => {};

    (
        $(#[$meta:meta])*
        $opt:ident = $value:expr,
        $level:expr,
        $rtyp:ty,
        $($tr:ty),*
    ) => {
```

The macro headers are easier to read when indented. I don't think there is any point of having a specific `unimplemented` matcher that does nothing.

Named fields would make these macros more clear, and you could use repetition:

```rust
    impl_ip_opts!{
        /// Don't reserve a port when binding with port number 0.
        ///
        /// C value type: `int`
        BindAddressNoPort {
            value: bindings::IP_BIND_ADDRESS_NO_PORT,
            ctype: bool,
            options: [ReadableOption, WritableOption]
        },
        BlockSource {
            // ..
        }
    }
```

All that being said, the RFL team sometimes prefer having the code written out than using macros. I could see this going either way :)